### PR TITLE
Change from Thread to Process

### PR DESCRIPTION
**Pull request by Kelvin-M**
## Description
It looks like using threads to download models in parallel is slower than Process.
For 11 * 2 Go, 

**Thread** 
```
#1 -  49.08039045333862 seconds - 2097152000                                                               
#5 - 67.21909880638123 seconds - 2097152000                                                               
#2 - 69.7494068145752 seconds - 2097152000                                                                
#4 - 73.38624572753906 seconds - 2097152000                                                               
#3 - 76.63423752784729 seconds - 2097152000                                                               
#6 - 78.72280073165894 seconds - 2097152000                                                               
#7 - 81.33725309371948 seconds - 2097152000                                                               
#11 - 81.71029543876648 seconds - 2097152000                                                              
#8 - 84.1978645324707 seconds - 2097152000                                                                
#9 - 84.2302668094635 seconds - 2097152000                                                                
#10 - 82.81430339813232 seconds - 2097152000  
```
**Process**
```
#1 - 12.778266668319702 seconds - 2097152000
#7 - 11.798902988433838 seconds - 2097152000
#2 - 13.3548104763031 seconds - 2097152000
#3 - 12.40976595878601 seconds - 2097152000
#4 - 11.725588083267212 seconds - 2097152000
#6 - 12.409850358963013 seconds - 2097152000
#5 - 13.262670040130615 seconds - 2097152000
#8 - 13.058980703353882 seconds - 2097152000
#10 - 11.18153691291809 seconds - 2097152000
#9 - 13.346837043762207 seconds - 2097152000
#11 - 12.924207925796509 seconds - 2097152000
```

## Closes issue(s)
None

## Companion PRs
None

## How to test / repro
None

## Screenshots / Trace
None

## Changes include
- [x] Migrate Thread to Process

## Checklist
None

## Other comments
None

### CODE REVIEW

**Comment by Kelvin-M** [comment]
Launching e2e tests !

**Review by AlexandrePicosson — APPROVED** [review]

**Comment by Kelvin-M** [comment]
It looks like we cannot launch processes under celery tasks :(

**Comment by Kelvin-M** [comment]
using `Process` from `billiard` (celery multiprocessing fork) works 🎉

**Comment by AurelienGasser** [comment]
@Kelvin-M Nice work! Could you please add a condition to bypass the Process creation in the case when there's only one model to fetch?

It would also be interesting to test it if the approach in solution 6 [here](https://stackoverflow.com/a/35819773/1370722) would be even more performant, although this can be done separately from this PR.

**Comment by Kelvin-M** [comment]
I would rather prefer to keep the same behavior as before. Moreover, using the same function for 1, 2 or n `in_models`, as it is a list, is preferable than handling each case. If we bypass the Process creation for one model, maybe for 2 models, Threads are faster and for n > 1000 asyncio/aiohttp is the best solution. So having Process for all, as it looks to perform well, is, imo, the good change.

For solutions proposed in the stackoverflow, it optimises RPS, which is not the case for us. I try an experiment with asyncio and it was clearly slower :(

File: backend/substrapp/tasks/tasks.py
Comment (samlesu) [inline backend/substrapp/tasks/tasks.py:241]: Why do you create a list of exceptions ?

You could create a list of errors directly as an Exception is already raised line 244.

It would be great to add also the return code value in the error message.

File: backend/substrapp/tasks/tasks.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:240]: Ah, you're right ! I missed this attribute !

File: backend/substrapp/tasks/tasks.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:227]: Yes, maybe :)  But like Aurélien suggestion, it's not the aim of this PR to change the current behavior !

File: backend/substrapp/tasks/tasks.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:241]: I keep the old behavior: raise an Exception with the list of exceptions in each parallel job.
Return code is useless as Process log already the python Traceback :)

Comment (samlesu) [reply backend/substrapp/tasks/tasks.py:241]: I don't think it makes sense to encapsulate the error messages in exceptions, but that's fine for me.

**Review by samlesu — APPROVED** [review]

File: backend/substrapp/tasks/tasks.py
Comment (AurelienGasser) [inline backend/substrapp/tasks/tasks.py:229]: @Kelvin-M just to be clear, my suggestion is to use a function call (not a thread) when there's only one model to download:


```suggestion
    if len(input_models) == 1:
        # Don't create sub-processes if there's only one model to download
        fetch_model(channel_name, tuple_type, authorized_types, input_model, directory)
        return

    models = []
```

Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:229]: Thank you :)
As answered [here](https://github.com/SubstraFoundation/substra-backend/pull/351#issuecomment-733540836), I would prefer to use the same solution for all cases than handle them with a special behavior for each (here 2, but we can do an analysis to know when it's better to use sequential calls or parallel calls) as it's not "slower" to spawn a Process

Comment (AurelienGasser) [reply backend/substrapp/tasks/tasks.py:229]: I just ran a quick benchark and process creation takes 6ms on my machine, nothing to argue about here :)

File: backend/substrapp/tasks/utils.py
Comment (AurelienGasser) [inline backend/substrapp/tasks/utils.py:42]: Nit: IMO it would be clearer if `auth` was a required parameter

**Review by AurelienGasser — APPROVED** [review]
LGTM

File: backend/substrapp/tasks/utils.py
Comment (Kelvin-M) [reply backend/substrapp/tasks/utils.py:42]: This modification was done to handle the particular issue of using `Process` to fetch all models in parallel which seems to mess the usage of Django db model to get OutGoing authentication properly.
For that, we need to get and give the authentication before launching the function in the `Process`, but we may want to keep the old behavior (get authentication in the function) for other call.

File: backend/substrapp/tasks/tasks.py
Comment (inalgnu) [inline backend/substrapp/tasks/tasks.py:240]: Just curious @Kelvin-M what kind of random results you get here? I think even we can merge this we should do further investigations in orderer to understand what is happening

Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:240]: Retrieving sometime a wrong password for a particular `node_id` instead of the good one :(

**Comment by Kelvin-M** [comment]
Do not merge right now, I will try to see if we can fix the auth problem without modifying the authentication code part !

**Pull request merged**
